specs/dwritecore/DWriteCore.md

```diff
@@ -1050,9 +1258,591 @@ int __cdecl wmain(int argc, WCHAR** argv)
+    HRESULT hr = m_dwriteFactory->TranslateColorGlyphRun(
+        baselineOrigin,
+        glyphRun,
+        glyphRunDescription,
+        desiredGlyphImageFormats,
+        measuringMode,
+        reinterpret_cast<DWRITE_MATRIX const*>(&transform),
+        /*colorPaletteIndex*/ 0,
+        /*out*/ & colorRunEnumerator
+    );
+
+    // Handle DWRITE_E_NOCOLOR, which is returned if the base glyph run does not
+    // have any color glyphs in the desired glyph image formats. In this case,
+    // just render the base glyph run as a monochrome glyph run.
+    if (hr == DWRITE_E_NOCOLOR)
+    {
+        m_deviceContext->DrawGlyphRun(
+            baselineOrigin,
+            glyphRun,
+            foregroundBrush,
+            measuringMode
+        );
+        return S_OK;
+    }
+
+    // Any other failure HRESULT from TranslateColorGlyphRun is an error.
+    RETURN_IF_FAILED(hr);
+
+    // Solid color brush to be lazily created if needed.
+    wil::com_ptr<ID2D1SolidColorBrush> solidBrush;
+
+    // Use the returned enumerator object to iterate over the color glyph runs.
+    for (;;)
+    {
+        // Advanced to the first or next run.
+        BOOL haveRun;
+        RETURN_IF_FAILED(colorRunEnumerator->MoveNext(&haveRun));
+        if (!haveRun)
+            break;
+
+        // Get a pointer to the color glyph run structure.
+        DWRITE_COLOR_GLYPH_RUN1 const* colorGlyphRun;
+        RETURN_IF_FAILED(colorRunEnumerator->GetCurrentRun(&colorGlyphRun));
+
+        // Determine the brush to use for this color glyph run.
+        ID2D1Brush* runBrush;
+        if (colorGlyphRun->paletteIndex == 0xFFFF)
+        {
+            // Special palette index meaning use the text foreground brush.
+            runBrush = foregroundBrush;
+        }
+        else
+        {
+            // Use the specified color from the font's color palette.
+            if (solidBrush == nullptr)
+            {
+                // Lazily create the solid color brush with the specified color.
+                RETURN_IF_FAILED(m_deviceContext->CreateSolidColorBrush(
+                    colorGlyphRun->runColor,
+                    &solidBrush
+                    ));
+            }
+            else
+            {
+                // We already created the brush, so just set its color.
+                solidBrush->SetColor(colorGlyphRun->runColor);
+            }
+
+            // use the solid color brush as the current run's brush.
+            runBrush = solidBrush.get();
+        }
+
+        // Branch depending on the run's glyph image format.
+        switch (colorGlyphRun->glyphImageFormat)
+        {
+        case DWRITE_GLYPH_IMAGE_FORMATS_NONE:
+            // do nothing
+            break;
+
+        case DWRITE_GLYPH_IMAGE_FORMATS_PNG:
+        case DWRITE_GLYPH_IMAGE_FORMATS_JPEG:
+        case DWRITE_GLYPH_IMAGE_FORMATS_TIFF:
+        case DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8:
+            m_deviceContext->DrawColorBitmapGlyphRun(
```

Comment:
These 3 methods share common arguments, such as baselineOrigin, glyphRun, measuringMode, and so on, but the order of these various arguments seems to differ across the methods. Ideally we should make sure the ordering of the common arguments is consistent as possible across the methods. For example, here baselineOrigin is second in the first method, but first in the other methods, and other common params have differing positions/orders in the other methods

Recommend all common parameters come first in these methods, then method-specific params come second. Unclear to me if the ordering differences here are intentional and have rationale or not.